File: src/ugv_teleop/src/keyboard_teleop_node.cpp
#include <algorithm>
#include <array>
#include <atomic>
#include <chrono>
#include <csignal>
#include <cstring>
#include <fcntl.h>
#include <fstream>
#include <functional>
#include <glob.h>
#include <iostream>
#include <memory>
#include <sstream>
#include <string>
#include <thread>
#include <unordered_map>
#include <vector>

#include <linux/input.h>
#include <sys/ioctl.h>
#include <sys/select.h>
#include <unistd.h>

#include "rcl_interfaces/msg/parameter_descriptor.hpp"
#include "rclcpp/rclcpp.hpp"
#include "ugv_teleop/keyboard_teleop_common.hpp"

namespace ugv_teleop
{

namespace
{
constexpr unsigned long k_evioc_grab = 0x40044590;

const std::unordered_map<int, std::string> k_key_code_to_char = {
  {16, "q"},
  {17, "w"},
  {18, "e"},
  {23, "i"},
  {24, "o"},
  {30, "a"},
  {31, "s"},
  {32, "d"},
  {37, "k"},
  {38, "l"},
  {45, "x"},
};

std::vector<std::string> discover_evdev_keyboard_devices(const std::string & explicit_device)
{
  if (!explicit_device.empty()) {
    if (access(explicit_device.c_str(), F_OK) == 0) {
      return {explicit_device};
    }
    return {};
  }

  std::vector<std::string> devices;
  glob_t glob_result;
  std::memset(&glob_result, 0, sizeof(glob_result));
  if (glob("/dev/input/by-path/*-event-kbd", 0, nullptr, &glob_result) == 0) {
    for (std::size_t i = 0; i < glob_result.gl_pathc; ++i) {
      devices.emplace_back(glob_result.gl_pathv[i]);
    }
  }
  globfree(&glob_result);
  std::sort(devices.begin(), devices.end());
  if (!devices.empty()) {
    return devices;
  }

  std::ifstream devices_file("/proc/bus/input/devices");
  if (!devices_file.is_open()) {
    return {};
  }

  std::vector<std::string> fallback;
  std::string line;
  while (std::getline(devices_file, line)) {
    const std::string prefix = "H: Handlers=";
    if (line.rfind(prefix, 0) != 0) {
      continue;
    }
    const std::string handlers = line.substr(prefix.size());
    if (handlers.find("kbd") == std::string::npos) {
      continue;
    }
    std::istringstream handler_stream(handlers);
    std::string token;
    while (handler_stream >> token) {
      if (token.rfind("event", 0) == 0) {
        fallback.push_back("/dev/input/" + token);
      }
    }
  }
  std::sort(fallback.begin(), fallback.end());
  fallback.erase(std::unique(fallback.begin(), fallback.end()), fallback.end());
  return fallback;
}

}

class KeyboardTeleopNode : public rclcpp::Node
{
public:
  KeyboardTeleopNode()
  : rclcpp::Node("keyboard_teleop_node")
  {
    common_params_ = declare_common_parameters(*this);
    core_ = create_core_from_params(common_params_);
    publisher_ = create_twist_publisher_adapter(*this, common_params_);

    rcl_interfaces::msg::ParameterDescriptor read_only_descriptor;
    read_only_descriptor.read_only = true;
    evdev_device_ = this->declare_parameter<std::string>(
      "evdev_device", "", read_only_descriptor);
    grab_device_ = this->declare_parameter<bool>(
      "grab_device", false, read_only_descriptor);
    evdev_poll_timeout_ = this->declare_parameter<double>(
      "evdev_poll_timeout", 0.005, read_only_descriptor);
    evdev_poll_timeout_ = std::max(0.001, evdev_poll_timeout_);

    const auto publish_period = std::chrono::duration<double>(
      1.0 / std::max(1.0, common_params_.publish_rate));
    publish_timer_ = this->create_wall_timer(
      publish_period,
      std::bind(&KeyboardTeleopNode::publish_timer_callback, this));

    const auto discovered = discover_evdev_keyboard_devices(evdev_device_);
    for (const auto & path : discovered) {
      if (access(path.c_str(), R_OK) == 0) {
        readable_devices_.push_back(path);
      }
    }
    if (readable_devices_.empty()) {
      throw std::runtime_error("No readable keyboard evdev device found. Please set evdev_device.");
    }

    keyboard_thread_ = std::thread(&KeyboardTeleopNode::keyboard_thread_main, this);
    RCLCPP_WARN(
      this->get_logger(),
      "Using legacy evdev keyboard backend. This mode may conflict with desktop input.");
  }

  ~KeyboardTeleopNode() override
  {
    shutdown();
  }

  void shutdown()
  {
    if (shutdown_done_.exchange(true)) {
      return;
    }
    stop_requested_.store(true);
    if (keyboard_thread_.joinable()) {
      keyboard_thread_.join();
    }
    core_->emergency_stop(std::chrono::steady_clock::now());
    if (rclcpp::ok()) {
      publisher_->publish_zero(core_.get());
    }
  }

private:
  void publish_timer_callback()
  {
    const auto command = core_->snapshot(std::chrono::steady_clock::now());
    publisher_->publish(command);
  }

  void keyboard_thread_main()
  {
    std::vector<int> fds;
    bool ctrl_pressed = false;

    try {
      for (const auto & path : readable_devices_) {
        const int fd = open(path.c_str(), O_RDONLY | O_NONBLOCK);
        if (fd < 0) {
          continue;
        }
        if (grab_device_) {
          if (ioctl(fd, k_evioc_grab, 1) < 0) {
            close(fd);
            continue;
          }
        }
        fds.push_back(fd);
      }

      if (fds.empty()) {
        stop_requested_.store(true);
        if (rclcpp::ok()) {
          rclcpp::shutdown();
        }
        return;
      }

      while (rclcpp::ok() && !stop_requested_.load()) {
        fd_set read_set;
        FD_ZERO(&read_set);
        int max_fd = -1;
        for (int fd : fds) {
          FD_SET(fd, &read_set);
          max_fd = std::max(max_fd, fd);
        }

        timeval timeout{};
        timeout.tv_sec = static_cast<long>(evdev_poll_timeout_);
        timeout.tv_usec = static_cast<long>((evdev_poll_timeout_ - timeout.tv_sec) * 1e6);
        const int ready = select(max_fd + 1, &read_set, nullptr, nullptr, &timeout);
        if (ready <= 0) {
          continue;
        }

        for (int fd : fds) {
          if (!FD_ISSET(fd, &read_set)) {
            continue;
          }
          std::array<input_event, 128> events{};
          const ssize_t bytes_read = read(fd, events.data(), sizeof(events));
          if (bytes_read <= 0) {
            continue;
          }
          const std::size_t event_count =
            static_cast<std::size_t>(bytes_read) / sizeof(input_event);
          const auto now = std::chrono::steady_clock::now();

          for (std::size_t i = 0; i < event_count; ++i) {
            const auto & event = events[i];
            if (event.type != EV_KEY) {
              continue;
            }

            if (event.code == KEY_LEFTCTRL || event.code == KEY_RIGHTCTRL) {
              ctrl_pressed = (event.value == 1 || event.value == 2);
              continue;
            }

            if (event.code == KEY_C && event.value == 1 && ctrl_pressed) {
              stop_requested_.store(true);
              if (rclcpp::ok()) {
                rclcpp::shutdown();
              }
              break;
            }

            const auto key_it = k_key_code_to_char.find(event.code);
            if (key_it == k_key_code_to_char.end()) {
              continue;
            }
            const std::string & key = key_it->second;
            if (KeyboardTeleopCore::is_move_key(key)) {
              if (event.value == 0) {
                core_->handle_key_release(key, now);
              } else if (event.value == 1 || event.value == 2) {
                core_->handle_key_press(key, now);
              }
            } else if (KeyboardTeleopCore::is_speed_key(key) && event.value == 1) {
              core_->handle_key_press(key, now);
            }
          }
        }
      }
    } catch (const std::exception & exception) {
      RCLCPP_ERROR(this->get_logger(), "Evdev keyboard thread failed: %s", exception.what());
    }

    core_->clear_move_keys(std::chrono::steady_clock::now());

    for (int fd : fds) {
      if (grab_device_) {
        (void)ioctl(fd, k_evioc_grab, 0);
      }
      close(fd);
    }
  }

  CommonTeleopParams common_params_;
  std::shared_ptr<KeyboardTeleopCore> core_;
  std::shared_ptr<TwistPublisherAdapter> publisher_;
  rclcpp::TimerBase::SharedPtr publish_timer_;

  std::string evdev_device_;
  bool grab_device_{false};
  double evdev_poll_timeout_{0.005};
  std::vector<std::string> readable_devices_;

  std::atomic<bool> stop_requested_{false};
  std::atomic<bool> shutdown_done_{false};
  std::thread keyboard_thread_;
};

}

int main(int argc, char ** argv)
{
  std::cout << ugv_teleop::k_common_keyboard_message << std::endl;

  rclcpp::init(argc, argv);
  int exit_code = 0;

  try {
    auto node = std::make_shared<ugv_teleop::KeyboardTeleopNode>();
    rclcpp::spin(node);
    node->shutdown();
  } catch (const std::exception & exception) {
    RCLCPP_ERROR(
      rclcpp::get_logger("keyboard_teleop_node"),
      "Unhandled exception: %s",
      exception.what());
    exit_code = 1;
  }

  if (rclcpp::ok()) {
    rclcpp::shutdown();
  }

  return exit_code;
}

File: src/ugv_teleop/ugv_teleop/keyboard_teleop_node.py
#!/usr/bin/env python3
import fcntl
import glob
import os
import select
import struct
import threading
import time

import rcl_interfaces.msg
import rclpy

from .keyboard_teleop_common import (
    COMMON_KEYBOARD_MESSAGE,
    create_core_from_params,
    create_twist_publisher_adapter,
    declare_common_parameters,
)
from .keyboard_teleop_core import MOVE_BINDINGS, SPEED_BINDINGS


EV_KEY = 0x01
KEY_LEFTCTRL = 29
KEY_RIGHTCTRL = 97
KEY_C = 46

KEY_CODE_TO_CHAR = {
    16: "q",
    17: "w",
    18: "e",
    23: "i",
    24: "o",
    30: "a",
    31: "s",
    32: "d",
    37: "k",
    38: "l",
    45: "x",
}

EVIOCGRAB = 0x40044590


def discover_evdev_keyboard_devices(explicit_device):
    if explicit_device:
        if os.path.exists(explicit_device):
            return [explicit_device]
        return []

    devices = sorted(glob.glob('/dev/input/by-path/*-event-kbd'))
    if devices:
        return devices

    fallback = []
    try:
        with open('/proc/bus/input/devices', 'r', encoding='utf-8') as fh:
            for line in fh:
                line = line.strip()
                if not line.startswith('H: Handlers='):
                    continue
                handlers = line.split('=', 1)[1].split()
                if 'kbd' not in handlers:
                    continue
                for handler in handlers:
                    if handler.startswith('event'):
                        fallback.append(f'/dev/input/{handler}')
    except OSError:
        return []
    return sorted(set(fallback))


class EvdevKeyboardThread(threading.Thread):
    def __init__(self, core, stop_event, device_paths, grab_device, poll_timeout):
        super().__init__(daemon=True)
        self.core = core
        self.stop_event = stop_event
        self.device_paths = device_paths
        self.grab_device = bool(grab_device)
        self.poll_timeout = max(0.001, float(poll_timeout))

    def run(self):
        event_format = 'llHHI'
        event_size = struct.calcsize(event_format)
        fd_to_path = {}
        buffers = {}
        ctrl_pressed = False

        try:
            for path in self.device_paths:
                fd = None
                try:
                    fd = os.open(path, os.O_RDONLY | os.O_NONBLOCK)
                    if self.grab_device:
                        fcntl.ioctl(fd, EVIOCGRAB, 1)
                except OSError:
                    try:
                        if fd is not None:
                            os.close(fd)
                    except Exception:
                        pass
                    continue
                fd_to_path[fd] = path
                buffers[fd] = b''

            if not fd_to_path:
                self.stop_event.set()
                return

            while rclpy.ok() and not self.stop_event.is_set():
                ready_fds, _, _ = select.select(list(fd_to_path.keys()), [], [], self.poll_timeout)
                if not ready_fds:
                    continue

                for fd in ready_fds:
                    try:
                        chunk = os.read(fd, event_size * 128)
                    except BlockingIOError:
                        continue
                    except OSError:
                        continue

                    if not chunk:
                        continue

                    buffer = buffers[fd] + chunk
                    event_count = len(buffer) // event_size
                    offset = 0

                    for _ in range(event_count):
                        _, _, event_type, code, value = struct.unpack_from(event_format, buffer, offset)
                        offset += event_size

                        if event_type != EV_KEY:
                            continue

                        if code in (KEY_LEFTCTRL, KEY_RIGHTCTRL):
                            ctrl_pressed = value in (1, 2)
                            continue

                        if code == KEY_C and value == 1 and ctrl_pressed:
                            self.stop_event.set()
                            break

                        key = KEY_CODE_TO_CHAR.get(code)
                        if key is None:
                            continue

                        timestamp = time.monotonic()
                        if key in MOVE_BINDINGS:
                            if value == 0:
                                self.core.handle_key_release(key, timestamp)
                            elif value in (1, 2):
                                self.core.handle_key_press(key, timestamp)
                        elif key in SPEED_BINDINGS and value == 1:
                            self.core.handle_key_press(key, timestamp)

                    buffers[fd] = buffer[offset:]
                    if self.stop_event.is_set():
                        break
        finally:
            self.core.clear_move_keys(time.monotonic())
            for fd in fd_to_path:
                try:
                    if self.grab_device:
                        fcntl.ioctl(fd, EVIOCGRAB, 0)
                except OSError:
                    pass
                try:
                    os.close(fd)
                except OSError:
                    pass


def main():
    rclpy.init()
    node = rclpy.create_node('keyboard_teleop_node')

    read_only_descriptor = rcl_interfaces.msg.ParameterDescriptor(read_only=True)
    params = declare_common_parameters(node)
    evdev_device = node.declare_parameter('evdev_device', '', read_only_descriptor).value
    grab_device = node.declare_parameter('grab_device', False, read_only_descriptor).value
    evdev_poll_timeout = node.declare_parameter('evdev_poll_timeout', 0.005, read_only_descriptor).value

    core = create_core_from_params(params)
    publisher = create_twist_publisher_adapter(node, params)
    publish_period = 1.0 / max(1.0, float(params['publish_rate']))
    node.create_timer(publish_period, lambda: publisher.publish(core.snapshot(time.monotonic())))

    device_paths = discover_evdev_keyboard_devices(str(evdev_device))
    readable_devices = [path for path in device_paths if os.access(path, os.R_OK)]
    if not readable_devices:
        node.get_logger().error('No readable keyboard evdev device found. Please set evdev_device.')
        node.destroy_node()
        rclpy.shutdown()
        return

    spinner = threading.Thread(target=rclpy.spin, args=(node,), daemon=True)
    spinner.start()

    stop_event = threading.Event()
    keyboard_thread = EvdevKeyboardThread(
        core=core,
        stop_event=stop_event,
        device_paths=readable_devices,
        grab_device=grab_device,
        poll_timeout=evdev_poll_timeout,
    )

    try:
        node.get_logger().warn('Using legacy evdev keyboard backend. This mode may conflict with desktop input and is not recommended for daily use.')
        print(COMMON_KEYBOARD_MESSAGE)
        keyboard_thread.start()

        while rclpy.ok() and not stop_event.is_set():
            time.sleep(0.05)
    except KeyboardInterrupt:
        pass
    finally:
        stop_event.set()
        if keyboard_thread.is_alive():
            keyboard_thread.join(timeout=1.0)
        core.emergency_stop(time.monotonic())
        publisher.publish_zero(core)
        node.destroy_node()
        rclpy.shutdown()
        spinner.join(timeout=1.0)


if __name__ == '__main__':
    main()
